### PR TITLE
Remove `rec` keyword

### LINKED ISSUE #36: Drop the `rec` keyword
Having to define recursive functions writing `fn rec` instead of `fn` feels a little bit uncomfortable. Not just because the language doesn't have a lint for that, but also because no other language (excluding OCaml and its family) requires the special tag for recursion.

The simplest solution would be to make all functions recursive by default. However that'd cause an overhead during execution because recursive functions need an extra step to be evaluated (our recursion execution is based on the [fixed-point combinator](https://en.wikipedia.org/wiki/Fixed-point_combinator)). 

A better solution would be to traverse the body of a function looking for the name of the function inside the AST. If the name is present, the function is recursive. Sadly, it is not as simple as that. We'd need to:
- [x] Implement the traversal stated above, but be careful when names are rebounded. For example, the following is a (badly written) non-recursive function:
    ```
    fn foo(x: Int) do 
        foo = 1
        x + foo
    end
    ```
- [x] If a function is recursive, we need it's return type because the type checker cannot infer it yet. If a function ends up being recursive and the return type is not there, then we'd need to throw an error asking for the return type. 
- [x] Add new tests for the last 2 items.
- [x] Update the tests and examples.

One of the main difficulties here is that all of this must be done before lowering to the MIR. So we might need a new stage before that lowering  to do these sort of checks.

### PR DESCRIPTION
**Pull request by pvdrz**
fixes #36. It was harder than I thought because it required to traverse the AST looking for recursive callings taking shadowing into account.

### CODE REVIEW

File: src/ast/analysis.rs
Comment (pvdrz) [inline src/ast/analysis.rs:10]: @DarkDrek could you give this a look? I'm worried that it might not work as intended and ends in some sort of ICE.

File: src/ast/analysis.rs
Comment (DarkDrek) [inline src/ast/analysis.rs:71]: `super_name()` doesn't do anything so it is just there for completeness and symmetry with the other methods?

File: src/ast/visitor.rs
Comment (DarkDrek) [inline src/ast/visitor.rs]: BinOp can be passed per value (size_of BinOP < sizeof reference)

File: src/ast/visitor.rs
Comment (DarkDrek) [inline src/ast/visitor.rs]: Instead of `&Vec` a slice reference is usually better `&[]`

File: src/ast/analysis.rs
Comment (DarkDrek) [reply src/ast/analysis.rs:10]: I looked at it (`RecursionChecker` and `NodeVisitor`) and gave a few comments why do you think it will give an ICE?

**Comment by DarkDrek** [comment]
This program doesn't type check `Err(Unbound("foo"))`

Edit: switching the if arms it does...?
```
fn foo(x: Int): Int do
    if true do
        fn foo(x: Int) do
            x + 1
        end
        foo(x)
    else
        foo(x)
    end
end

foo
```

File: src/ast/analysis.rs
Comment (pvdrz) [reply src/ast/analysis.rs:71]: Yes, I have to document the visitor trait to make that clear. The idea is that you should always call `super_<foo>`  in your implementation of `visit_<foo>`

File: src/ast/visitor.rs
Comment (pvdrz) [reply src/ast/visitor.rs]: Is it there any other reason for that besides coercion? If someone needs coercion there probably they're misusing the trait.

**Comment by pvdrz** [comment]
> This program doesn't type check `Err(Unbound("foo"))`
> 
> Edit: switching the if arms it does...?
> 
> ```
> fn foo(x: Int): Int do
>     if true do
>         fn foo(x: Int) do
>             x + 1
>         end
>         foo(x)
>     else
>         foo(x)
>     end
> end
> 
> foo
> ```

Oh wow, of course. I'm not taking conditionals into account at all! I'll fix it and add this as a test.

Update: I added your example as a test case (a fail version and a pass version) and updated the visiting of the `RecursionChecker` to work with conditionals.

You made me realize that `Block`s aren't just "sequences of `Node`s". They are the only place where a binding can actually have an effect.

File: src/ast/analysis.rs
Comment (pvdrz) [reply src/ast/analysis.rs:10]: This is just some gut feeling so take it with a grain of salt.

If for some weird reason `RecursionChecker` ends up saying that a function is not recursive and the function is. That means that it won't get wrapped in a `Term::Fix`. Which also means that the place where the function calls itself will have an unbounded name. 

This could cause the type checker to panic when it's checking the body of the recursive function, or in the best case scenario to actually say that the name of the function is unbounded.

But in a total disaster scenario where the type checker doesn't realize that the name is unbounded. Then the lowering to the `LIR` will panic, or worse it will try to execute the program and panic because there aren't enough indices for all the variables.

I don't have any hard reasons to believe this will fail. It is only that the notion of scoping inside the `RecursionChecker` has to coincide with the notion of scoping in the lower representations.

File: src/ast/analysis.rs
Comment (pvdrz) [reply src/ast/analysis.rs:71]: I added some docs to the vistor trait so it is clearer how it works

File: src/ast/visitor.rs
Comment (pvdrz) [reply src/ast/visitor.rs]: Oh there is an extra reference (of course, vecs have an slice inside them). I'll change this.

File: src/ast/visitor.rs
Comment (pvdrz) [reply src/ast/visitor.rs]: Thanks! I've done it now :)

**Review by DarkDrek — APPROVED** [review]
A lot of new things and complexity but I like the approach. Go ahead and merge.

PS: You need to give yourself a approval too 😆

**Comment by pvdrz** [comment]
Heh I had to rebase because of the new changes but now it is ready. I'll wait for CI just in case

**Comment by christianpoveda** [comment]
Heh I had to rebase because of the new changes but now it is ready. I'll wait for CI just in case

**Pull request merged**
